lib/SIL/SILFunctionType.cpp

```diff
@@ -1242,6 +1029,510 @@ class DestructureInputs {
 
 } // end anonymous namespace
 
+/// Collects the differentiability parameters of the given original function
+/// type in `diffParams`.
+static void
+getDifferentiabilityParameters(SILFunctionType *originalFnTy,
+                               IndexSubset *parameterIndices,
+                               SmallVectorImpl<SILParameterInfo> &diffParams) {
+  // Returns true if `index` is a differentiability parameter index.
+  auto isDiffParamIndex = [&](unsigned index) -> bool {
+    return index < parameterIndices->getCapacity() &&
+           parameterIndices->contains(index);
+  };
+  // Calculate differentiability parameter infos.
+  for (auto valueAndIndex : enumerate(originalFnTy->getParameters()))
+    if (isDiffParamIndex(valueAndIndex.index()))
+      diffParams.push_back(valueAndIndex.value());
+}
+
+/// Collects the semantic results of the given function type in
+/// `originalResults`. The semantic results are formal results followed by
+/// `inout` parameters, in type order.
+// TODO(TF-983): Generalize to support multiple `inout` parameters. The current
```

Comment:
Note: the singular `inoutParam` and `isWrtInoutParameter` parameters are hacky, but it's a NFC refactoring from existing code. TF-983 tracks improving these.

We should support multiple `inout` parameters. Maybe we need a `SmallVectorImpl<AutoDiffInoutParameterInfo>`:
```cpp
// fileprivate type
struct AutoDiffInoutParameterInfo {
  SILParameterInfo inoutParam;
  bool isWrt;
};
```